be/src/runtime/memory_scratch_sink.cpp

```diff
@@ -48,13 +49,24 @@ MemoryScratchSink::MemoryScratchSink(const RowDescriptor& row_desc, const std::v
 
 MemoryScratchSink::~MemoryScratchSink() {}
 
+void MemoryScratchSink::convert_to_slot_types_and_ids() {
+    for (auto* tuple_desc : _row_desc.tuple_descriptors()) {
+        auto& slots = tuple_desc->slots();
+        for (auto slot : slots) {
+            _slot_types.push_back(&slot->type());
+            _slot_ids.push_back(slot->id());
+        }
+    }
+}
+
 Status MemoryScratchSink::prepare_exprs(RuntimeState* state) {
     // From the thrift expressions create the real exprs.
     RETURN_IF_ERROR(Expr::create_expr_trees(state->obj_pool(), _t_output_expr, &_output_expr_ctxs));
     // Prepare the exprs to run.
     RETURN_IF_ERROR(Expr::prepare(_output_expr_ctxs, state, _row_desc, _expr_mem_tracker.get()));
     // generate the arrow schema
     RETURN_IF_ERROR(convert_to_arrow_schema(_row_desc, &_arrow_schema));
+    convert_to_slot_types_and_ids();
```

Comment:
RETURN_IF_ERROR

Reply:
it always succeeds and return type is void